### PR TITLE
rustdoc: change `.src-line-numbers > span` to `.src-line-numbers > a`

### PR DESCRIPTION
**Pull request by notriddle**
Example: https://notriddle.com/notriddle-rustdoc-demos/line-anchors/test_dingus/fn.test.html

This allows people to treat them like real links, such as right-click to copy URL, and makes the line numbers in a scraped example work at all, when before this commit was added, they had the clickable pointer cursor but did not actually do anything when clicked.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
With this change, when another line is selected, the original one keeps the right border:

image

Just to explain a bit more what I did:

 1. Click on a example line number.
 2. When on the source code page, click on another line.
 3. The border of the "original" line is still orange.

When switched back to `<span>`, it doesn't keep it, so I suppose it's a new bug. Once fixed, can you add a test to ensure that it doesn't happen please?

File: src/librustdoc/html/static/js/source-script.js [hunk 1]
```diff
@@ -211,6 +216,12 @@ const handleSourceHighlight = (function() {
     };
 }());
 
+const addSourceLineAnchorsHref = () => {
+    onEachLazy(document.querySelectorAll(".src-line-numbers a"), e => {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/source-script.js:220]: This is unneeded. If there is an ID, it automatically links to itself if there is no `href`.

File: src/librustdoc/html/static/js/scrape-examples.js [hunk 2]
```diff
@@ -79,6 +80,19 @@
 
         // Start with the first example in view
         scrollToLoc(example, locs[0][0]);
+
+        const addSourceLineAnchorsHref = () => {
+            onEachLazy(document.querySelectorAll(".src-line-numbers a"), e => {
+                e.href = linkPage + "#" + e.innerText;
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/scrape-examples.js:86]: On small pages, it's fine. However, it's a very bad ideas as soon as you start to have more than 5 functions with examples which comes pretty quickly. Either generate directly the URLs (which would increase the size) or don't change it at all.

File: src/librustdoc/html/static/js/source-script.js (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/source-script.js:220]: What is the right clicking supposed to do?

File: src/librustdoc/html/static/js/scrape-examples.js (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/scrape-examples.js]: Big question: is it worth it? It'll increase the page size quite a lot.

File: src/librustdoc/html/static/js/source-script.js (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/source-script.js]: Oh indeed. It doesn't propose to copy a link, what a shame... Then I guess it comes to this: is it better to generate `href` through JS or in the HTML directly?

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/source-script.js]: Let's see once we have results to think about. ;)

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I think it's acceptable. r=me once you have fixed the merge conflict.

File: src/librustdoc/html/sources.rs [hunk 3]
```diff
@@ -270,31 +270,32 @@ pub(crate) fn print_src(
...
         }
-        SourceContext::Embedded { offset, needs_expansion } => {
+        SourceContext::Embedded { url, offset, needs_expansion } => {
             extra =
                 if needs_expansion { Some(r#"<span class="expand">&varr;</span>"#) } else { None };
-            for line in 1..=lines {
-                writeln!(line_numbers, "<span>{0}</span>", line + offset)
+            for line_number in 1..=lines {
+                let line = line_number + offset;
+                writeln!(line_numbers, "<a href=\"{root_path}{url}#{line}\">{line}</a>")
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/sources.rs:294]: This one is still bugging me: as soon as this feature is enabled by default, the size will increase like crazy... I'm really not sure it's worth it to generate links here.

Comment (Nemo157) [reply src/librustdoc/html/sources.rs:294]: I assume the `doc_bytes` benchmarks are raw file sizes? Seems like this should compress really well and not affect the transferred size much.

**Comment by GuillaumeGomez** [comment]
Then I suggest this: let's only keep this change for the source code pages and remove it for the examples as it is where it will crazily increase.

**Review by GuillaumeGomez — APPROVED** [review]

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors r+ rollup

**Pull request merged**